File: src/main/java/io/binac/leetcode/RecoverBinarySearchTree.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * Two elements of a binary search tree (BST) are swapped by mistake.
 * <p>
 * <p>Recover the tree without changing its structure.
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: [1,3,null,null,2]
 *
 *        1
 *       /
 *      3
 *       \
 *        2
 *
 *     Output: [3,1,null,null,2]
 *
 *        3
 *       /
 *      1
 *       \
 *        2
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: [3,1,4,null,null,2]
 *
 *        3
 *       / \
 *      1   4
 *         /
 *        2
 *
 *     Output: [2,1,4,null,null,3]
 *
 *        2
 *       / \
 *      1   4
 *         /
 *        3
 * </blockquote></pre>
 * Follow up:
 * <p>
 * <p>A solution using O(n) space is pretty straight forward.
 * <p>Could you devise a constant space solution?
 */
public class RecoverBinarySearchTree {
    public static class Solution1 {
        private void inOrder(TreeNode root, List<TreeNode> result) {
            if (root == null) return;
            inOrder(root.left, result);
            result.add(root);
            inOrder(root.right, result);
        }

        public void recoverTree(TreeNode root) {
            List<TreeNode> nodes = new ArrayList<>();
            inOrder(root, nodes);
            int i = 0, j = nodes.size() - 1;
            while (i < nodes.size() - 1 && nodes.get(i).val < nodes.get(i + 1).val) ++i;
            while (j > 0 && nodes.get(j).val > nodes.get(j - 1).val) --j;
            if (i < nodes.size() - 1 && j > 0) {
                int tmp = nodes.get(i).val;
                nodes.get(i).val = nodes.get(j).val;
                nodes.get(j).val = tmp;
            }
        }
    }
}

File: python/leetcodepy/recover_binary_search_tree.py
"""
Two elements of a binary search tree (BST) are swapped by mistake.

Recover the tree without changing its structure.

Example 1:

Input: [1,3,null,null,2]

   1
  /
 3
  \
   2

Output: [3,1,null,null,2]

   3
  /
 1
  \
   2

Example 2:

Input: [3,1,4,null,null,2]

  3
 / \
1   4
   /
  2

Output: [2,1,4,null,null,3]

  2
 / \
1   4
   /
  3

Follow up:

A solution using O(n) space is pretty straight forward.
Could you devise a constant space solution?
"""
from typing import List, Optional

from .utils import TreeNode


class Solution1:
    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        first, second = [None], [None]
        self.in_order(root, [None], first, second)
        if first[0] is not None and second[0] is not None:
            first[0].val, second[0].val = second[0].val, first[0].val

    def in_order(self, root: TreeNode, pre: List[Optional[TreeNode]],
                 first: List[Optional[TreeNode]], second: List[Optional[TreeNode]]):
        if root is None:
            return
        self.in_order(root.left, pre, first, second)
        if pre[0] is not None and pre[0].val >= root.val:
            if first[0] is None:
                first[0] = pre[0]
            second[0] = root
        pre[0] = root
        self.in_order(root.right, pre, first, second)
